lib/spack/spack/architecture.py

```diff
@@ -56,49 +56,133 @@
 attributes front_os and back_os. The operating system as described earlier,
 will be responsible for compiler detection.
 """
+import functools
 import inspect
 
+import six
+
+import llnl.util.cpu as cpu
 import llnl.util.tty as tty
 from llnl.util.lang import memoized, list_modules, key_ordering
 
 import spack.compiler
 import spack.paths
 import spack.error as serr
+import spack.version
 from spack.util.naming import mod_to_class
 from spack.util.spack_yaml import syaml_dict
 
 
 class NoPlatformError(serr.SpackError):
-
     def __init__(self):
         super(NoPlatformError, self).__init__(
             "Could not determine a platform for this machine.")
 
 
-@key_ordering
-class Target(object):
-    """ Target is the processor of the host machine.
-        The host machine may have different front-end and back-end targets,
-        especially if it is a Cray machine. The target will have a name and
-        also the module_name (e.g craype-compiler). Targets will also
-        recognize which platform they came from using the set_platform method.
-        Targets will have compiler finding strategies
+def _ensure_other_is_target(method):
+    """Decorator to be used in dunder methods taking a single argument to
+    ensure that the argument is an instance of ``Target`` too.
     """
+    @functools.wraps(method)
+    def _impl(self, other):
+        if isinstance(other, six.string_types):
+            other = Target(other)
+
+        if not isinstance(other, Target):
+            return NotImplemented
+
+        return method(self, other)
 
+    return _impl
+
+
+class Target(object):
     def __init__(self, name, module_name=None):
-        self.name = name  # case of cray "ivybridge" but if it's x86_64
-        self.module_name = module_name  # craype-ivybridge
+        """Target models the processor of the current architecture.
+
+        Args:
+            name (str or MicroArchitecture):micro-architecture of the
+                target
+            module_name (str): optional module name to get access to the
+                current target. This is typically used on machines
+                like Cray (e.g. craype-compiler)
+        """
+        if not isinstance(name, cpu.MicroArchitecture):
+            name = cpu.targets.get(
+                name, cpu.generic_microarchitecture(name)
+            )
+        self.micro_architecture = name
```

Comment:
why is it `microarchitecture` above, but `micro_architecture` here?  I think we should stick with `microarchitecture` (since that is the word)

Reply:
Fixed